### PR TITLE
Connectedk8s - RBAC: Add "arc "mode in Guard for using OBO service and deprecate --app-id & --app-secret params

### PR DESCRIPTION
**Pull request by vineeth-thumma**
**Arc RBAC changes for Connectedk8s extension**
- Add "arc "mode in Guard for using 1P apps for authN/authZ
- Deprecate --app-id & --app-secret params
-  Add warning to use latest kubelogin with PoP token support
- Bump the release version

E2e run with CLI version `1.5.5`:
 https://dev.azure.com/ClusterConfigurationAgent/ClusterConfigurationAgent/_build/results?buildId=26677&view=results
 

PR to bump CLI version in E2e: https://github.com/azure-core/ClusterConfigurationAgent/pull/1979

MicrosoftTeams-image (3)
MicrosoftTeams-image (2)
MicrosoftTeams-image (1)

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [x] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
Work In Progress - Do Not Merge

File: src/connectedk8s/azext_connectedk8s/custom.py [hunk 1]
```diff
@@ -1487,8 +1483,10 @@ def enable_features(cmd, client, resource_group_name, cluster_name, features, ku
         cmd_helm_upgrade.extend(["--kube-context", kube_context])
     if enable_azure_rbac:
         cmd_helm_upgrade.extend(["--set", "systemDefaultValues.guard.enabled=true"])
-        cmd_helm_upgrade.extend(["--set", "systemDefaultValues.guard.clientId={}".format(azrbac_client_id)])
-        cmd_helm_upgrade.extend(["--set", "systemDefaultValues.guard.clientSecret={}".format(azrbac_client_secret)])
+        # Setting the default authnMode mode as "arc" for guard. This mode used Arc RBAC 1P apps for authN/authZ.
```
Comment (rharpavat) [inline src/connectedk8s/azext_connectedk8s/custom.py:1486]: Nit: `uses`.

File: src/connectedk8s/azext_connectedk8s/custom.py [hunk 2]
```diff
@@ -1487,8 +1483,10 @@ def enable_features(cmd, client, resource_group_name, cluster_name, features, ku
...
+        # This mode in guard uses PoP token based auth.
```
Comment (rharpavat) [inline src/connectedk8s/azext_connectedk8s/custom.py:1487]: Nit: can combine this comment line with previous for brevity

File: src/connectedk8s/azext_connectedk8s/custom.py [hunk 3]
```diff
@@ -1487,8 +1483,10 @@ def enable_features(cmd, client, resource_group_name, cluster_name, features, ku
...
+        cmd_helm_upgrade.extend(["--set", "systemDefaultValues.guard.authnMode=arc"])
+        logger.warning("Please use the latest kubelogin version which has support for generating PoP token(s) needed by guard running in 'arc' authN mode.")
```
Comment (rharpavat) [inline src/connectedk8s/azext_connectedk8s/custom.py:1489]: Do we need to add a specific minimum version here? Not sure but can discuss

File: src/connectedk8s/azext_connectedk8s/tests/latest/test_connectedk8s_scenario.py [hunk 4]
```diff
@@ -150,9 +150,9 @@ class Connectedk8sScenarioTest(LiveScenarioTest):
     @live_only()
     @ResourceGroupPreparer(name_prefix='conk8stest', location=CONFIG['location'], random_name_length=16)
     def test_connect(self,resource_group):
-        
+
```
Comment (rharpavat) [inline src/connectedk8s/azext_connectedk8s/tests/latest/test_connectedk8s_scenario.py:153]: Nit: can we remove this empty line?

File: src/connectedk8s/HISTORY.rst [hunk 5]
```diff
@@ -3,6 +3,10 @@
 Release History
 ===============
 
+1.5.5
+++++++
+* Setting guard default authnMode to 'arc' when RBAC is enabled.
```
Comment (julienstroheker) [inline src/connectedk8s/HISTORY.rst:8]: Please be more specific here. `guard` is implementation details.

Comment (julienstroheker) [reply src/connectedk8s/HISTORY.rst:8]: You can describe all the changes we made such as the deprecation of the flags etc...

File: src/connectedk8s/azext_connectedk8s/_params.py [hunk 6]
```diff
@@ -80,8 +80,8 @@ def load_arguments(self, _):
         c.argument('kube_config', options_list=['--kube-config'], help='Path to the kube config file.')
         c.argument('kube_context', options_list=['--kube-context'], help='Kubconfig context from current machine.')
         c.argument('features', features_types, options_list=['--features'], help='Space-separated list of features you want to enable.')
-        c.argument('azrbac_client_id', options_list=['--app-id'], arg_group='Azure RBAC', help='Application ID for enabling Azure RBAC. Specify when enabling azure-rbac.')
-        c.argument('azrbac_client_secret', options_list=['--app-secret'], arg_group='Azure RBAC', help='Application secret for enabling Azure RBAC. Specify when enabling azure-rbac.')
+        c.argument('azrbac_client_id', options_list=['--app-id'], arg_group='Azure RBAC', help='To be deprecated soon. Application ID for enabling Azure RBAC.', deprecate_info=c.deprecate(hide=True))
```
Comment (julienstroheker) [inline src/connectedk8s/azext_connectedk8s/_params.py:83]: IMO we don't need to add `To be deprecated soon`, when doing a `--help` this should show up by it self thanks to the deprecate_info.

File: src/connectedk8s/azext_connectedk8s/custom.py [hunk 7]
```diff
@@ -1487,8 +1483,10 @@ def enable_features(cmd, client, resource_group_name, cluster_name, features, ku
         cmd_helm_upgrade.extend(["--kube-context", kube_context])
     if enable_azure_rbac:
         cmd_helm_upgrade.extend(["--set", "systemDefaultValues.guard.enabled=true"])
-        cmd_helm_upgrade.extend(["--set", "systemDefaultValues.guard.clientId={}".format(azrbac_client_id)])
-        cmd_helm_upgrade.extend(["--set", "systemDefaultValues.guard.clientSecret={}".format(azrbac_client_secret)])
+        # Setting the default authnMode mode as "arc" for guard. This mode used Arc RBAC 1P apps for authN/authZ.
+        # This mode in guard uses PoP token based auth.
+        cmd_helm_upgrade.extend(["--set", "systemDefaultValues.guard.authnMode=arc"])
```
Comment (julienstroheker) [inline src/connectedk8s/azext_connectedk8s/custom.py:1488]: Can this be defaulted on the chart side ?

File: src/connectedk8s/azext_connectedk8s/custom.py (hunk 3 above)
Comment (julienstroheker) [reply src/connectedk8s/azext_connectedk8s/custom.py:1489]: +1, let's be specific and asking for version "0.0.32" ? and up @rharpavat please confirm the version here

**Review by julienstroheker — CHANGES_REQUESTED** [review]
PTAL

File: src/connectedk8s/azext_connectedk8s/custom.py (hunk 3 above)
Comment (rharpavat) [reply src/connectedk8s/azext_connectedk8s/custom.py:1489]: Confirming that minimum version required for kubelogin is `v0.0.32`.

File: src/connectedk8s/azext_connectedk8s/custom.py (hunk 7 above)
Comment (vineeth-thumma) [reply src/connectedk8s/azext_connectedk8s/custom.py:1488]: If we default this on the chart size, will there be any implications for customers who are using RBAC in 3P mode and have agent auto upgrade to true? Can we do it in phases?
@julienstroheker

**Comment by vineeth-thumma** [comment]
@yonzhan @yanzhudd @zhoxing-ms for reviews

**Review by julienstroheker — APPROVED** [review]

**Comment by yanzhudd** [comment]
Hi @vineeth-thumma, please rerun tests in live mode and upload the recoding files if the recording files exist.

File: src/connectedk8s/azext_connectedk8s/custom.py [hunk 8]
```diff
@@ -1487,8 +1483,9 @@ def enable_features(cmd, client, resource_group_name, cluster_name, features, ku
         cmd_helm_upgrade.extend(["--kube-context", kube_context])
     if enable_azure_rbac:
         cmd_helm_upgrade.extend(["--set", "systemDefaultValues.guard.enabled=true"])
-        cmd_helm_upgrade.extend(["--set", "systemDefaultValues.guard.clientId={}".format(azrbac_client_id)])
-        cmd_helm_upgrade.extend(["--set", "systemDefaultValues.guard.clientSecret={}".format(azrbac_client_secret)])
+        # Setting the default authnMode mode as "arc" for guard. This mode uses PoP token based auth. and Arc RBAC 1P apps for authN/authZ.
+        cmd_helm_upgrade.extend(["--set", "systemDefaultValues.guard.authnMode=arc"])
+        logger.warning("Please use the kubelogin version v0.0.32 or higher which has support for generating PoP token(s) which is needed by guard running in 'arc' authN mode.")
```
Comment (rharpavat) [inline src/connectedk8s/azext_connectedk8s/custom.py:1488]: Nit: maybe reword this as:
"Please use kubelogin version v0.0.32 or higher, which has support for generating PoP token(s). This is required by guard when running in 'arc' authN mode."

File: src/connectedk8s/HISTORY.rst [hunk 9]
```diff
@@ -3,6 +3,12 @@
 Release History
 ===============
 
+1.5.5
+++++++
+* Deprecate 'app-id' and 'app-secret' RBAC params.
```
Comment (zhoxing-ms) [inline src/connectedk8s/HISTORY.rst:8]:
```suggestion
* Deprecate '--app-id' and '--app-secret' RBAC params.
```

File: src/connectedk8s/azext_connectedk8s/custom.py (hunk 8 above)
Comment (zhoxing-ms) [reply src/connectedk8s/azext_connectedk8s/custom.py:1488]: If you change the default behavior, you need to confirm the impact of the breaking change on the customers. 
We usually recommend implementing the breaking change in the breaking change window (**Ignite** and **Build** sprint), and before that, we suggest recommending customers to change their behavior through notification messages (warning log)

File: src/connectedk8s/azext_connectedk8s/custom.py (hunk 7 above)
Comment (zhoxing-ms) [reply src/connectedk8s/azext_connectedk8s/custom.py:1487]: If you change the default behavior, you need to confirm the impact of the breaking change on the customers.
We usually recommend implementing the breaking change in the breaking change window (Ignite and Build sprint), and before that, we suggest recommending customers to change their behavior through notification messages (warning log)

**Comment by vineeth-thumma** [comment]
Adding @atchutbarli for review

File: src/connectedk8s/azext_connectedk8s/custom.py (hunk 7 above)
Comment (vineeth-thumma) [reply src/connectedk8s/azext_connectedk8s/custom.py:1487]: Leaving the defaults as-is for now and will revisit later

**Comment by vineeth-thumma** [comment]
@yonzhan @yanzhudd @zhoxing-ms Can one of you please add @atchutbarli as a reviewer? (I'm unable to add). We need an approval from him as their team owns the extensiion.

**Review by atchutbarli — APPROVED** [review]

**Review by rharpavat — APPROVED** [review]

**Review by zhoxing-ms — APPROVED** [review]
